File: latest/week2/workshop5_m1/W05_M01_demo.py
"""W05-M01: High-Dimensional Data — Python — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, seaborn as sns, os
os.makedirs("output",exist_ok=True); np.random.seed(42); n=2000
fico=np.random.normal(700,50,n).clip(500,850); rate=12-0.01*fico+np.random.normal(0,1,n)
amount=np.random.uniform(5000,50000,n); tier=np.digitize(fico,[0,620,680,720,760,900])
cartype=np.random.choice(["New","Used"],n,p=[0.6,0.4])
ecar=pd.DataFrame({"FICO":fico,"Rate":rate,"Amount":amount,"Tier":tier,"CarType":cartype})

# 1. Multi-encoded scatter
fig,ax=plt.subplots(figsize=(8,6))
for ct,marker in [("New","o"),("Used","s")]:
    m=ecar["CarType"]==ct
    sc=ax.scatter(ecar.loc[m,"FICO"],ecar.loc[m,"Rate"],c=ecar.loc[m,"Tier"],s=ecar.loc[m,"Amount"]/300,
        marker=marker,alpha=0.4,cmap="viridis",label=ct,vmin=1,vmax=5)
plt.colorbar(sc,label="Tier"); ax.legend(title="Car Type",fontsize=7)
ax.set_xlabel("FICO"); ax.set_ylabel("Rate (%)"); ax.set_title("5 Variables, 1 Scatter",fontweight="bold")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/multi_scatter_py.png",dpi=300); plt.close()

# 2. Hexbin
fig,ax=plt.subplots(figsize=(7,5))
hb=ax.hexbin(ecar["FICO"],ecar["Rate"],gridsize=30,cmap="viridis",mincnt=1)
plt.colorbar(hb,label="Count"); ax.set_title("Hexbin: Aggregated Scatter",fontweight="bold")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/hexbin_py.png",dpi=300); plt.close()

# 3. Jointplot with marginals
g=sns.jointplot(data=ecar,x="FICO",y="Rate",hue="CarType",
    palette={"New":"#1565C0","Used":"#E53935"},kind="scatter",height=6,
    joint_kws={"s":10,"alpha":0.3})
g.fig.suptitle("Jointplot: Scatter + Marginal Histograms",fontweight="bold",y=1.02)
plt.savefig("output/jointplot_py.png",dpi=300,bbox_inches="tight"); plt.close()

# 4. 2D KDE
fig,ax=plt.subplots(figsize=(7,5))
sns.kdeplot(data=ecar,x="FICO",y="Rate",fill=True,cmap="Blues",levels=10,ax=ax)
ax.set_title("2D KDE: Density Contours",fontweight="bold")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/kde2d_py.png",dpi=300); plt.close()

print("All W05-M01 Python plots saved")

File: latest/week2/workshop5_m1/W05_M01_demo.R
# ============================================================
# Workshop 5 — Module 1: High-Dimensional Data
# Integrated R Script — UNYT Tirana
# ============================================================
library(tidyverse); library(patchwork); dir.create("output",showWarnings=FALSE)

# Simulated e-Car-like data for demo
set.seed(42); n <- 2000
ecar <- tibble(
  FICO = rnorm(n, 700, 50) |> pmax(500) |> pmin(850),
  Rate = 12 - 0.01*FICO + rnorm(n, 0, 1),
  Amount = runif(n, 5000, 50000),
  Tier = cut(FICO, breaks=c(0,620,680,720,760,900), labels=1:5),
  CarType = sample(c("New","Used"), n, replace=TRUE, prob=c(0.6,0.4)))

# 1. MULTI-ENCODED SCATTER — 5 variables
p_multi <- ggplot(ecar, aes(x=FICO, y=Rate, color=as.numeric(Tier), size=Amount, shape=CarType)) +
  geom_point(alpha=0.4) +
  scale_color_viridis_c(name="Tier") +
  scale_size(range=c(1,6), name="Amount ($)") +
  scale_shape_manual(values=c(New=16, Used=17)) +
  theme_minimal() + labs(title="5 Variables on 1 Scatter", x="FICO", y="Rate (%)")
ggsave("output/multi_scatter.png", p_multi, width=8, height=6, dpi=300)

# 2. HEXBIN — for large n
p_hex <- ggplot(ecar, aes(x=FICO, y=Rate)) +
  geom_hex(bins=30) + scale_fill_viridis_c() +
  theme_minimal() + labs(title="Hexbin: Aggregated Scatter for Large n")
ggsave("output/hexbin.png", p_hex, width=7, height=5, dpi=300)

# 3. MARGINAL DISTRIBUTIONS
library(ggExtra)
p_base <- ggplot(ecar, aes(x=FICO, y=Rate, color=CarType)) +
  geom_point(alpha=0.2, size=0.5) +
  scale_color_manual(values=c(New="#1565C0", Used="#E53935")) +
  theme_minimal(base_size=9) + theme(legend.position="bottom")
p_marg <- ggMarginal(p_base, type="histogram", groupColour=TRUE, groupFill=TRUE)
ggsave("output/marginal.png", p_marg, width=7, height=6, dpi=300)

# 4. OVERLOAD DEMO — too many encodings
p_over <- ecar |> mutate(PriceBin=cut(Amount,5)) |>
  ggplot(aes(x=FICO,y=Rate,color=PriceBin,size=Amount,shape=CarType)) +
  geom_point(alpha=0.3) + theme_minimal(base_size=8) +
  labs(title="Overloaded: 5 colour bins × 2 shapes × continuous size\n(hard to read)")
ggsave("output/overload.png", p_over, width=8, height=6, dpi=300)

cat("\n── All W05-M01 R plots saved ──\n")
